packages/report-server/src/reportBuilder/context/extractContextProps.ts

```diff
@@ -0,0 +1,57 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import isPlainObject from 'lodash.isplainobject';
+
+import { getUniqueEntries } from '@tupaia/utils';
+
+import { contextConfig, TransformParams, TransformParser } from '../transform';
+import { ContextProp } from './types';
+
+const extractContextPropsFromExpressions = (expressions: string[]) => {
+  const parser = new TransformParser();
+  const functions = expressions
+    .filter(parser.isCalculation)
+    .map(calcExpression => {
+      return parser.getFunctions(calcExpression);
+    })
+    .flat();
+
+  const contextProps = Object.entries(contextConfig)
+    .filter(([fnName]) => functions.includes(fnName))
+    .map(([, config]) => config.contextProps)
+    .flat();
+
+  return getUniqueEntries(contextProps);
+};
+
+export const extractContextProps = (transform: unknown): ContextProp[] => {
+  if (!Array.isArray(transform)) {
+    return [];
+  }
```

Comment:
Yes that's what I thought too! An approach would be:

1. Expand `this.config.transform` to an array of objects (eg by inlining aliases to objects)
2. Validate the expanded transform
3. Use the expanded transform to find all functions that use context
4. Use the expanded transform to build the output

So validation would happen first, and the rest of the functions would expected a validated & expanded `Transform` object as an input. This will make function extraction simpler, since the `transform` in its input will have a more specific schema.

We use a similar approach in `indicators` (we expand the aggregation input)

Reply:
I can submit a PR implementing the above at a later stage, but yeah for now let's proceed with what we have